optuna/samplers/_tpe/sampler.py

```diff
@@ -835,9 +835,18 @@ def _feasible(trial: FrozenTrial) -> bool:
 
     loo_mat = ~np.eye(pareto_sols.shape[0], dtype=bool)  # Leave-one-out bool matrix.
     contribs = np.zeros(n_below_feasible, dtype=float)
-    contribs[on_front] = hv - np.array(
-        [compute_hypervolume(pareto_sols[loo], ref_point, assume_pareto=True) for loo in loo_mat]
-    )
+    if len(study.directions) <= 3:
+        contribs[on_front] = [
+            hv - compute_hypervolume(pareto_sols[loo], ref_point, assume_pareto=True)
+            for loo in loo_mat
+        ]
+    else:
+        hvs = []
+        for i in range(pareto_sols.shape[0]):
+            limited_sols = np.maximum(pareto_sols[loo_mat[i, :]], pareto_sols[i, :])
+            intersec_hv = compute_hypervolume(limited_sols, ref_point)
+            hvs.append(np.prod(ref_point - pareto_sols[i, :]) - intersec_hv)
+        contribs[on_front] = hvs
```

Comment:
This version is a bit quicker:)

```suggestion
        contribs[on_front] = np.prod(ref_point - pareto_sols, axis=-1)
        limited_sols = np.maximum(pareto_sols, pareto_sols[:, np.newaxis])
        contribs[on_front] -= [
            compute_hypervolume(limited_sols[i, loo], ref_point)
            for i, loo in enumerate(loo_mat)
        ]
```